File: de2/software/de2_zap/Card.c
#include "Card.h"

void initCards(CardCtrl* cardCtrl) {
    int i;
    for (i = 0; i < MAX_CARDS; i++) {
        cardCtrl->deck[i] = i + 1;
        cardCtrl->discard[i] = 0;
    }
    for (i = 0; i < MAX_CARDS; i++)
    {
        int j = i + rand() % (MAX_CARDS - i);
        if (j >= MAX_CARDS)
            printf("Randomizer exceeded maximum index\n");
        int temp = cardCtrl->deck[j];
        cardCtrl->deck[j] = cardCtrl->deck[i];
        cardCtrl->deck[i] = temp;
    }
    printf("Printing deck order:\n");
    for (i = 0; i < MAX_CARDS; i++) {
        printf("%d, ", cardCtrl->deck[i]);
    }
    printf("\n");
    cardCtrl->deckIndex = 0;
    cardCtrl->discardIndex = 0;
}

Card testDrawCard(CardCtrl* cardCtrl) {
    cardCtrl->deckIndex++;
    return cardCtrl->deck[cardCtrl->deckIndex - 1];
}

const char * cardToString(Card id) {
    if (id > 0 && id <= 2)
        return "Schofield";
    else if (id <= 3)
        return "RevCarabine";
    else if (id <= 4)
        return "Winchester";
    else if (id <= 6)
        return "Volcanic ";
    else if (id <= 7)
        return "Schofield ";
    else if (id <= 8)
        return "Remington ";
    else if (id <= 33)
        return "Zap!";
    else if (id <= 45)
        return "Missed!";
    else if (id <= 51)
        return "Beer";
    else if (id <= 55)
        return "Panic";
    else if (id <= 59)
        return "Cat Balou";
    else if (id <= 61)
        return "Stagecoach";
    else if (id <= 62)
        return "Wells Fargo";
    else if (id <= 63)
        return "Gatling";
    else if (id <= 66)
        return "Duel";
    else if (id <= 68)
        return "Aliens";
    else if (id <= 70)
        return "Gen. Store";
    else if (id <= 71)
        return "Saloon";
    else if (id <= 74)
        return "Space Jail";
    else if (id <= 75)
        return "Dynamite";
    else if (id <= 77)
        return "Barrel";
    else if (id <= 78)
        return "Scope";
    else if (id <= 80)
        return "Mustang";
    else
        return "Invalid";
}

File: android/src/com/zap/main/Card.java
package com.zap.main;

public class Card {
	public final int cid;
	public final String name;
	public final char border;
	public final char number;
	public final char suit;
	public final boolean zap;
	public final boolean missed;
	public final int life;
	public final boolean forceDiscard;
	public final int draw;
	public final boolean onePlayer;
	public final boolean allPlayers;
	public final boolean onePlayerReachable;
	public final int onePlayerFixed;
	public final String image;
	//space duel, indians, general store
	//space jail, dynamite, barrel, scope, mustang
	Card(int cid, String name, char border, char number, char suit, boolean zap,
			boolean missed, int life, boolean forceDiscard, int draw, boolean onePlayer,
			boolean allPlayers, boolean onePlayerReachable, int onePlayerFixed, String image) {
		this.cid = cid;
		this.name = name;
		this.border = border;
		this.number = number;
		this.suit = suit;
		this.zap = zap;
		this.missed = missed;
		this.life = life;
		this.forceDiscard = forceDiscard;
		this.draw = draw;
		this.onePlayer = onePlayer;
		this.allPlayers = allPlayers;
		this.onePlayerReachable = onePlayerReachable;
		this.onePlayerFixed = onePlayerFixed;
		this.image = image;
	}

	public boolean isGunCard() {
		if (border == 'L' && onePlayerFixed != 0) {
			return true;
		} else {
			return false;
        }
    }

    public boolean isBlue() {
        return border == 'L';
    }

    public boolean isRegular() {
        return border == 'R';
    }

    public boolean selfTarget() {
        return isGunCard() || isMissed() || isBeer() || isStageCoach()
                || isWellsFargo() || isBarrel() || isMustang() || isScope()
                || isDynamite();
    }

    public boolean allPlayersIncSelf() {
        return isSaloon() || isGeneralStore();
    }

    public boolean allPlayersNotSelf() {
        return isGatling() || isAliens();
    }

    public boolean onePlayerNotSelf() {
        return isBang() || isPanic() || isCatBalou() || isDuel() || isJail();
    }

    public boolean isBang() {
        return name.equals("Zap!");
    }

    public boolean isDuel() {
        return name.equals("Duel");
    }

    public boolean isAliens() {
        return name.equals("Aliens");
    }

    public boolean isGeneralStore() {
        return name.equals("General Store");
    }

    public boolean isMissed() {
        return name.equals("Missed");
    }

    public boolean isBeer() {
        return name.equals("Beer");
    }

    public boolean isStageCoach() {
        return name.equals("Stagecoach");
    }

    public boolean isWellsFargo() {
        return name.equals("Wells Fargo");
    }

    public boolean isBarrel() {
        return name.equals("Barrel");
    }

    public boolean isMustang() {
        return name.equals("Mustang");
    }

    public boolean isScope() {
        return name.equals("Scope");
    }

    public boolean isSaloon() {
        return name.equals("Saloon");
    }

    public boolean isDynamite() {
        return name.equals("Dynamite");
    }

    public boolean isGatling() {
        return name.equals("Gatling");
    }

    public boolean isPanic() {
        return name.equals("Panic");
    }

    public boolean isCatBalou() {
        return name.equals("Cat Balou");
    }

    public boolean isJail() {
        return name.equals("Space Jail");
    }
}
